### PR TITLE
[TechDocs CLI] adding cypress testing

### PR DESCRIPTION
**Pull request by emmaindal**
## Hey, I just made a Pull Request!

Closes: #8105 

This PR adds more testing to the TechDocs CLI, both testing TechDocs live preview in Backstage and MkDocs.

Note that the tests are not running as a part of any github workflow at the moment. So one need to manually run the tests.

Also note that the example project used is located in another repository, my suggestion is that we move the example project (or duplicate if needed) in to the techdocs-cli package to make the testing a bit more smooth. @backstage/techdocs-core If you agree - let me know and I can move that in as a part of this PR. 

Screenshot 2022-02-15 at 14 06 07

Screenshot 2022-02-15 at 14 06 21

#### :heavy_check_mark: Checklist

- [ ] A changeset describing the change and affected packages. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#creating-changesets))
- [ ] Added or updated documentation
- [ ] Tests for new functionality and regression tests for bug fixes
- [ ] Screenshots attached (for UI changes)
- [ ] All your commits have a `Signed-off-by` line in the message. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#developer-certificate-of-origin))

### CODE REVIEW

**Review by Rugvip — APPROVED** [review]
🟢 to unblock the yarn.lock change

File: packages/techdocs-cli/README.md
Comment (bodilb) [inline packages/techdocs-cli/README.md:45]:
```suggestion
For the purpose of local development, we have created an [example documentation project](https://github.com/backstage/techdocs-container/tree/main/mock-docs) which is shipped with the [techdocs-container](https://github.com/backstage/techdocs-container) repository. You are of course also free to create your own local test site - all it takes is a `docs/index.md` and an `mkdocs.yml` in a directory.
```

File: packages/techdocs-cli/README.md
Comment (bodilb) [inline packages/techdocs-cli/README.md:43]:
```suggestion
### Using an example docs project
```
For consistency (running, using, testing)

File: packages/techdocs-cli/README.md
Comment (bodilb) [inline packages/techdocs-cli/README.md:70]:
```suggestion
#### Running Cypress tests
```

File: packages/techdocs-cli/README.md
Comment (bodilb) [inline packages/techdocs-cli/README.md:74]:
```suggestion
Running cypress tests requires you to run the CLI locally against our example docs project.

Start by making sure that you have the example project locally on your computer:
```

File: packages/techdocs-cli/README.md
Comment (bodilb) [inline packages/techdocs-cli/README.md:80]:
```suggestion
Run the local version of techdocs-cli against the example docs project:
```

File: packages/techdocs-cli/README.md
Comment (bodilb) [inline packages/techdocs-cli/README.md:94]:
```suggestion
Run the cypress tests:
```

**Review by bodilb — APPROVED** [review]
Minor edits suggested for improved readability and consistency.

**Comment by emmaindal** [comment]
Added visual regression tests which let us run snapshot test and catch differences on the page. 


**Note:** One limitation with this is that the snapshots does not currently include the shadow dom, which makes it impossible for us to test the rendering of the content in Backstage but thats something we hopefully can add in the future. [Ref: 
 [https://github.com/cypress-io/cypress/issues/8843](https://github.com/cypress-io/cypress/issues/8843)]

Screenshot 2022-02-17 at 11 17 54

**Comment by emmaindal** [comment]
I'll look into the large yarn.lock changes, that wasn't expected!

File: packages/techdocs-cli/README.md
Comment (camilaibs) [inline packages/techdocs-cli/README.md:85]: We have a script that builds and runs the CLI locally in one go:
```sh
yarn techdocs-cli serve
```

File: packages/techdocs-cli/cypress/integration/backstage_serve.js
Comment (camilaibs) [inline packages/techdocs-cli/cypress/integration/backstage_serve.js:40]: Do you think it's a good idea not to rely on `mkdocs-material` classes and adding data-testids to these elements would be safer to prevent these tests from breaking?

**Review by camilaibs — APPROVED** [review]
Added minor comments, I'm looking forward to seeing this merged, it will surely help us out!

File: packages/techdocs-cli/cypress/integration/backstage_serve.js
Comment (emmaindal) [reply packages/techdocs-cli/cypress/integration/backstage_serve.js:40]: My thinking around it was - if the classes changes - we want the test to break 😬 Or? As we may rely on it in the Reader and therefore want to make sure we catch the update. But maybe worth checking both actually? What do you think @camilaibs?

File: packages/techdocs-cli/README.md
Comment (emmaindal) [reply packages/techdocs-cli/README.md:85]: good point, in this case we want to run the cli with another example docs, therefore building separately. But I had added the prepack script to the `cypress:open` script which has now disappeared so will move that back. And with that we don't need this step. Thanks for pointing out! 

Another question - do you think I should move/duplicate the example docs project in here instead? that makes it easier to just run one test command which will do everything we need?

File: packages/techdocs-cli/cypress/integration/backstage_serve.js
Comment (camilaibs) [reply packages/techdocs-cli/cypress/integration/backstage_serve.js:40]: Classes are generated by `mkdocs`, right? If so, will the tests only work for a Reader implementation that uses `mkdocs`? When [this](https://github.com/backstage/backstage/pull/9385) pull request is merged and people use a different Reader implementation what would you advise them to use for testing? Maybe if they use the same test ids the tests will continue to work regardless of the Reader chosen and the CSS used by it.

File: packages/techdocs-cli/README.md
Comment (camilaibs) [reply packages/techdocs-cli/README.md:85]: Yes, I found myself using the same `mock-docs/index.md` example when I was updating `mkdocs-material`. Maybe we can add the contents of `mock-docs/index.md` to a markdown file inside the Backstage `/docs` folder, on an examples page so we can preview the Backstage `/docs` folder and navigate to this new page in the test. What do you think?

File: packages/techdocs-cli/cypress/integration/backstage_serve.js
Comment (emmaindal) [reply packages/techdocs-cli/cypress/integration/backstage_serve.js:40]: > different Reader implementation

Then the embedded app needs to be updated anyways, as its currently using the one Reader that exist. So if/when that happens we can update the tests here too!

File: packages/techdocs-cli/README.md
Comment (emmaindal) [reply packages/techdocs-cli/README.md:85]: Yeah, the /docs documentation is pretty different as its the public docs thats used for (backstage.io) and that microsite uses docusaurus. So my suggestion is to add a docs folder to this package (techdocs-cli) if we want to move it.

File: packages/techdocs-cli/cypress/integration/backstage_serve.js
Comment (camilaibs) [reply packages/techdocs-cli/cypress/integration/backstage_serve.js:40]: You're right, so we can update the tests later, I agree with you!

File: packages/techdocs-cli/README.md
Comment (emmaindal) [reply packages/techdocs-cli/README.md:80]: I moved the docs in this PR, see https://github.com/backstage/backstage/pull/9546/commits/77a13948c15ec3b67b72596fb90a0605a15369a7, which also could make the instructions a bit more clear in the README I think https://github.com/backstage/backstage/pull/9546/commits/a6352f90eaf5b1a1377d39b0d2ed8097dd98d5e2

**Comment by emmaindal** [comment]
@backstage/maintainers can we quickly get your 👀  on this again!

**Comment by freben** [comment]
We'll have to add a changeset for techdocs-cli, right?

**Review by freben — APPROVED** [review]
Approved, but changeset should be added I think.

**Comment by emmaindal** [comment]
@freben I thought I don't need a changeset when its its just docs and tests changes?

**Comment by freben** [comment]
Yeah but this PR also changes package.json dependencies, README changes, pack script ... lots of stuff that actually affect the published package

**Comment by emmaindal** [comment]
Ok, got it, I'll add a changeset! Thanks!

**Pull request merged**
